### ISSUE TITLE
Why using a VAE instead of a AE?

### ISSUE THREAD

**Issue by francois-rozet** [cloneofsimo/vqgan-training#2]
Hello @cloneofsimo, very cool work! I have been working for quite some time on diffusion/latent diffusion stuff. I never understood why people use VAEs instead of simple AEs.

To my understanding the only reason why VAEs are used is to regularize the distribution of latents, the target being zero mean, unit variance. However using VAEs leads to a tradeoff between this regularization and the reconstruction. It is therefore necessary to include weights and additional loss terms (perceptual, discriminator, ...) which are hard to tune.

Instead, in recent experiments, I have used a simple AE with a very simple loss ($L_2(D(E(x)), x)$ or $L_1$), and it worked really well. The trick was to **saturate the output of the encoder**, to guarantee a latent distribution close to the standard Gaussian. In practice, I used a softclip between -5 and 5.

$$softclip(x) = \frac{x}{1 + \frac{|x|}{5}}$$

However, I have never tried this on natural images. I would be very curious to know if this works as well in this case.

**Comment by charchit7** [comment]
Hey, @francois-rozet nice idea. Was wondering how did you come up with the idea of using softclip?

**Comment by francois-rozet** [comment]
I was just looking for functions that have a bounded co-domain, but are not saturating extremely fast like $\tanh$, because it would lead to almost-zero gradients. One class of such functions are

$$ x \mapsto \frac{x}{\big( 1 + |x|^n \big)^\frac{1}{n}} $$

with $n \geq 1$. I chose $n = 1$ because it saturates very slowly.

**Comment by charchit7** [comment]
Got it, thanks.

**Comment by Oguzhanercan** [comment]
if your purpose is reconstruction, AE is a good choice but when you want to generate something from a random variable (usually a vector sampled from Gaussian distribution) the latent space of a is meaningless. The output of encoder at AE is only meaningful when you store it, you cannot randomly sample a vector and get a result. When it comes to VAE, conditional probability comes to the table. The KL divergence at VAE makes the distribution of latent space gaussian, so you can randomly sample from it and generate images.

**Comment by francois-rozet** [comment]
Plain VAEs are terrible at generating images from random noise. That is why people are using diffusion models to generate in the latent space. The only reason people are still using the KL in the training (at a very low weight) is to regularize the latent space towards a zero mean, unit variance distribution, compatible with the diffusion model. Using a saturating function has the same goal, while being simpler.

**Comment by Oguzhanercan** [comment]
Of course plain vae is not cool, but the key concept is matching the distribution with standard normal distribution (zero mean, unit variance). with perceptual loss, ssim, gan losses or other architectures like  β-VAEs, VQ-VAEs, we achieve high quality images. Since diffusion model's accuracy bounded with the generation quality of VAE (because modern diffusion&flow models are operating at latent space of VAE), we cannot say that VAEs are terrible at generating images, but we can say that sampling from distribution is hard. Denoising process at LDM is basically noise sampler for decoder of VAE (while being simpler).

KL guarantees smooth, probabilistic latent distribution (so you can add conditions like text or classes),  I do not think that your softclip solution guarantees that the latent space is distributed well between -5 and 5. For n=2, It could be like clustering distribution, around -5 and +5, so if you sample values around 0, decoder would produce meaningless outputs. More basically, there is no prior. KL guarantees that because it quantifies the additional information required to achieve desired distribution, which is N(0,I), not the difference or something like that.

**Comment by francois-rozet** [comment]
> KL guarantees that because it quantifies the additional information required to achieve the desired distribution, which is N(0,I).

I disagree. If you use a high enough KL weight for the latent distribution to actually be Gaussian, the VAE will be terrible at reconstruction. As you said, practitioners are using special losses (perceptual, adversarial, ...) and most importantly very small KL weights, to prevent over regularized latents. As a result the latent space distribution is highly non Gaussian with means and standard deviations all over the place (~4-7 of standard deviation for Stable Diffusion's autoencoder if I remember correctly?). We might as well remove the "variational" in the name at this point... It is just a regularized AE with adversarial reconstruction losses.

> I do not think that your softclip solution guarantees that the latent space is distributed well between -5 and 5

Because we are performing compression, it is in the interest of the auto-encoder to use the whole latent domain. Otherwise it would be compressing more than it needs, and have a worse reconstruction. Therefore, using a saturating function, we actually incite the network to be well spread, while the KL punishes it.

> It could be like clustering distribution, around -5 and +5

The softclip name might be misleading here. It is not clipping values. For the softclip to produce +4.99 it would need an input of ~500, which is highly unlikely. Besides, if the network manages to perform clustering in the latent space (without hurting reconstruction), it would lead to a distribution that is faster to sample for the diffusion model. Distributions with sharp modes/very structured/on a manifold, can be sampled in less steps.

In practice in my experiments, my AEs learn immediately to not use the values close to the bounds of the softclip as these values are highly saturated and hard to convey information with. Instead, the network  uses the range [-3, 3], which corresponds to the almost linear part of the soft-clip. All of my AEs so far had a zero latent mean, and a variance around 0.7, without any KL.

**Comment by Oguzhanercan** [comment]
If the only purpose of VAE is compressing in some way, then we can just resize images. Lets think that, we are working with 256x256x3 images (imagenet). Our latent space is  4x32x32 (flux or sd3.5 for 256x256 images), scale it with bits -> 256x256x3x8/4/32/32/32 = 12. So instead of using vae, you can just resize 256x256 -> 64x64 and you will have more compression rate :)

**Comment by francois-rozet** [comment]
An auto-encoder is able to compress much more than a simple resize, at fixed reconstruction loss. Note that people are moving to AEs for LDMs (see https://arxiv.org/abs/2410.10733).
